From 5563d11914a8d332eef4f398ca5fecf91a055026 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Thu, 10 Jul 2014 01:08:58 -0400
Subject: [acpi] pci_root: Clear driver_data before failing enumeration

Message-id: <20140710010858.22146.46539.stgit@gir.stowe>
Patchwork-id: 85641
O-Subject: [RHEL7.1 PATCH 102/105] ACPI / PCI root: Clear driver_data before failing enumeration
Bugzilla: 1114228
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1114228
Upstream Status: f516bde5122422889398d10e8e746d18860bd42e

commit f516bde5122422889398d10e8e746d18860bd42e
Author: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
Date:   Thu Nov 14 00:54:17 2013 +0100

    ACPI / PCI root: Clear driver_data before failing enumeration

    If a PCI host bridge cannot be enumerated due to an error in
    pci_acpi_scan_root(), its ACPI device object's driver_data field
    has to be cleared by acpi_pci_root_add() before freeing the
    object pointed to by that field, or some later acpi_pci_find_root()
    checks that should fail may succeed and cause quite a bit of
    confusion to ensue.

    Fix acpi_pci_root_add() to clear device->driver_data before
    returning an error code as appropriate.

    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
    Acked-by: Toshi Kani <toshi.kani@hp.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/acpi/pci_root.c b/drivers/acpi/pci_root.c
index b490ec8..20360e4 100644
--- a/drivers/acpi/pci_root.c
+++ b/drivers/acpi/pci_root.c
@@ -578,6 +578,7 @@ static int acpi_pci_root_add(struct acpi_device *device,
   dev_err(&device->dev,
    "Bus %04x:%02x not present in PCI namespace\n",
    root->segment, (unsigned int)root->secondary.start);
+  device->driver_data = NULL;
   result = -ENODEV;
   goto end;
  }
-- 
1.7.1